hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java

```diff
@@ -827,26 +828,34 @@ protected double scale(double min, double max, double value) {
    */
   static class MoveCostFunction extends CostFunction {
     private static final String MOVE_COST_KEY = "hbase.master.balancer.stochastic.moveCost";
+    private static final String MOVE_COST_OFFPEAK_KEY = "hbase.master.balancer.stochastic.moveCost.offpeak";
     private static final String MAX_MOVES_PERCENT_KEY =
         "hbase.master.balancer.stochastic.maxMovePercent";
-    private static final float DEFAULT_MOVE_COST = 7;
+    @VisibleForTesting
+    protected static final float DEFAULT_MOVE_COST = 7;
+    protected static final float DEFAULT_MOVE_COST_OFFPEAK = 3;
     private static final int DEFAULT_MAX_MOVES = 600;
     private static final float DEFAULT_MAX_MOVE_PERCENT = 0.25f;
 
     private final float maxMovesPercent;
+    private final Configuration conf;
 
     MoveCostFunction(Configuration conf) {
       super(conf);
-
-      // Move cost multiplier should be the same cost or higher than the rest of the costs to ensure
-      // that large benefits are need to overcome the cost of a move.
-      this.setMultiplier(conf.getFloat(MOVE_COST_KEY, DEFAULT_MOVE_COST));
+      this.conf = conf;
       // What percent of the number of regions a single run of the balancer can move.
       maxMovesPercent = conf.getFloat(MAX_MOVES_PERCENT_KEY, DEFAULT_MAX_MOVE_PERCENT);
     }
 
     @Override
     protected double cost() {
+      // Move cost multiplier should be the same cost or higher than the rest of the costs to ensure
+      // that large benefits are need to overcome the cost of a move.
+      if (OffPeakHours.getInstance(conf).isOffPeakHour()) {
+        this.setMultiplier(conf.getFloat(MOVE_COST_OFFPEAK_KEY, DEFAULT_MOVE_COST_OFFPEAK));
+      } else {
+        this.setMultiplier(conf.getFloat(MOVE_COST_KEY, DEFAULT_MOVE_COST));
+      }
```

Comment:
Why not keep this logic in `MoveCostFunction` constructor?

Reply:
The blancer only be initialized once on start, but we want the multiplier changed at offpeak hours.
Thanks.

Reply:
Good.

Reply:
@bsglz I just realized that whenever we update offpeak config dynamically, `MoveCostFunction` constructor is called so we should better have this logic in constructor only.

Check this out:
```
  @Override
  public void onConfigurationChange(Configuration conf) {
    setConf(conf);
  }

  @Override
  public synchronized void setConf(Configuration conf) {
    super.setConf(conf);
...
...
    costFunctions = new ArrayList<>();
    costFunctions.add(new RegionCountSkewCostFunction(conf));
    costFunctions.add(new PrimaryRegionCountSkewCostFunction(conf));
    costFunctions.add(new MoveCostFunction(conf));
    costFunctions.add(localityCost);
...
...
```
Hence, basically we can remove `this.conf = conf;` from this PR and move this if condition to constructor.
When we try to update the config dynamically for SLB, `onConfigurationChange()` is executed and that will reassign entire constFunctions list.

Reply:
@virajjasani Good question.

I glimpsed and I guess what @bsglz meant is every time calling the `MoveCostFunction::cost()` function against the same `MoveCostFunction` object should get a different value per the config and the current time. Since this `MoveCostFunction` object is constructed once, the multiplier should be updated here. We can not move the whole if condition to constructor. Updating the `config` object dynamically is another use case (which should also work). Correct me if I'm wrong @bsglz .

That said, to remove `this.conf = conf` dependency, we can get two multipliers from config, and call `setMultiplier()` accordingly here in `cost()` by checking `isOffPeakHour()`.

Reply:
Yeah, this makes sense. +1